src/match/controllers/update_match_controller.py

```diff
@@ -0,0 +1,121 @@
+import datetime
+import json
+
+from api.utils import failure_response
+from api.utils import success_response
+from location.models import Location
+from match import match_status
+from match.models import Match
+from rest_framework import status
+
+
+class UpdateMatchController:
+    def __init__(self, match_id, user, data, serializer):
+        self._match_id = match_id
+        self._user = user
+        self._data = data
+        self._serializer = serializer
+
+    def process(self):
+        # First, find the match and any fields to be updated
+        self._match = Match.objects.filter(id=self._match_id)
+        if not self._match:
+            return failure_response("Match does not exist.")
+        self._match = self._match[0]
+        proposed_meeting_times = self._data.get("proposed_meeting_times")
+        proposed_locations = self._data.get("proposed_locations")
+        meeting_location_id = self._data.get("meeting_location")
+        meeting_time = self._data.get("meeting_time")
+
+        # Next, modify the attributes that may have changed:
+        # Proposed Meeting Times
+        self._modify_attribute("proposed_meeting_times", proposed_meeting_times)
+        # Meeting Time
+        if meeting_time is not None:
+            self._update_meeting_time(meeting_time)
+        # Proposed Locations
+        if proposed_locations is not None:
+            for loc_id in proposed_locations:
+                new_location = Location.objects.filter(id=loc_id)
+                if not new_location:
+                    return failure_response(f"Location id {loc_id} does not exist.")
+                if new_location[0] not in self._match.proposed_locations.all():
+                    self._match.proposed_locations.add(new_location[0])
+            for existing_location in self._match.proposed_locations.all():
+                if existing_location.id not in proposed_locations:
+                    self._match.proposed_locations.remove(existing_location)
+        # Meeting Location
+        if meeting_location_id is not None:
+            new_meeting_location = Location.objects.filter(id=meeting_location_id)
+            if not new_meeting_location:
+                return failure_response(
+                    f"Location with id {meeting_location_id} does not exist."
+                )
+            self._modify_attribute("meeting_location", new_meeting_location[0])
+
+        # Based on what changed, check for status conflicts
+        if proposed_meeting_times is not None and proposed_locations is not None:
```

Comment:
I would check these BEFORE modifying the match because then you'll have the issue of "uh oh you can't do this" but you already did it and now your match's proposed meeting times or locations are in a messed up